### ISSUE TITLE
graphql.schema.idl.SchemaDirectiveWiring has a design flaw

### ISSUE THREAD

**Issue by bbakerman** [graphql-java/graphql-java#1535]
`graphql.schema.idl.SchemaDirectiveWiring` is where people can mutate objects and mostly important mutate fields and their data fetcher.

It calls back in a specific order, arguments then fields then  object.

`graphql.schema.idl.SchemaDirectiveWiring#onArgument`
`graphql.schema.idl.SchemaDirectiveWiring#onField`
`graphql.schema.idl.SchemaDirectiveWiring#onObject`

How the design flaw is that it calls back ONLY for each directive on the entity

So for example given

```
type ObjType {
     field( arg1 : String @length(max=20), arg2 : Int @max(value:100) ) : Book
}
```

the `onField` and `onObject` will never be called  because neither element has a directive on it.  However if some code want to find all the directives on the arguments of the field then it cant (easily) today  because there is no `onField` callback even though the sub entitiy has directives

Even if there are ZERO directives one could imagine wanting a callback such that code is generated BECAUSE there is zero directives

My thoughts at the moment is to have a callback per entity and not per dfirective per entity and callback EVEN when there is zero directives.

We can do this 3 ways

* add a `onFieldBuilt` callback method that is called after the existing
* add a whole new callback interface SPI for this and not piggy back the current one (not a fan)
* break SPI and steal back the well named `onField` / `onObject` and change the behaviour so you get ALL directives (zero or more) in the call

I am torn here as to which is best.  Breaking SPI is never a great idea but the naming and shape is the best way to do it this in retrospect

Having the extra `onFieldBuilt` callback will confuse people in the long term because it shows the design mistake and we pay for it forever.

**Comment by bbakerman** [comment]
As a work around to the "parent element does not get a  callback" you will need to put in a dummy directive.  So for example to get access to the @length and @max arguments directives in the example below

```
type ObjType {
     field( arg1 : String @length(max=20), arg2 : Int @max(value:100) ) : Book
}
```

you would do this to get a `onField` callback

```
type ObjType {
     field( arg1 : String @length(max=20), arg2 : Int @max(value:100) ) : Book @dummy
}
```

or

```
@dummy
type ObjType {
     field( arg1 : String @length(max=20), arg2 : Int @max(value:100) ) : Book 
}
```

to get a `onObject` callback

**Comment by bbakerman** [comment]
This is now fixed in 13.0
